src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmCodeGenerationServiceImpl.java

```diff
@@ -0,0 +1,202 @@
+		final List<GermplasmCodingResult> assignCodesResultsList = new ArrayList<>();
+		final boolean startNumberSpecified = setting.getStartNumber() != null;
+		Integer startNumber = setting.getStartNumber();
+
+		// Call this method to check first if the name settings are valid.
+		this.germplasmNamingService.getNextNameInSequence(setting);
+
+		for (final Integer gid : gids) {
+			// Increment start number of succeeding germplasm processed based on initial start # specified, if any
+			if (startNumberSpecified) {
+				setting.setStartNumber(startNumber++);
+			}
+			final String nameWithSequence = this.germplasmNamingService.generateNextNameAndIncrementSequence(setting);
+			final GermplasmCodingResult result = this.applyGroupName(gid, nameType, nameWithSequence);
+			assignCodesResultsList.add(result);
+		}
+		return assignCodesResultsList;
+	}
+
+	protected GermplasmCodingResult applyGroupName(final Integer gid,
+		final GermplasmNameTypeDTO nameType, final String generatedCodeName) {
+
+		final GermplasmCodingResult result = new GermplasmCodingResult();
+		result.setGid(gid);
+
+		final Germplasm germplasm = this.germplasmService.getGermplasmByGID(gid);
+
+		if (germplasm.getMgid() == null || germplasm.getMgid() == 0) {
+			result.addMessage(String.format(GERMPLASM_NOT_PART_OF_MANAGEMENT_GROUP, germplasm.getGid()));
+			return result;
+		}
+
+		final List<Germplasm> groupMembers =
+			this.germplasmGroupingService.getDescendantGroupMembers(germplasm.getGid(), germplasm.getMgid());
+		groupMembers.add(0, germplasm);
+
+		for (final Germplasm member : groupMembers) {
+			this.addName(member, generatedCodeName, nameType, result);
+		}
+
+		return result;
+
+	}
+
+	private void addName(final Germplasm germplasm, final String groupName, final GermplasmNameTypeDTO nameType,
+		final GermplasmCodingResult result) {
+
+		final List<Name> currentNames = germplasm.getNames();
+
+		Name existingNameOfGivenType = null;
+		if (!currentNames.isEmpty() && nameType != null) {
+			for (final Name name : currentNames) {
+				if (nameType.getId().equals(name.getTypeId())) {
+					existingNameOfGivenType = name;
+					break;
+				}
+			}
+		}
+
+		if (existingNameOfGivenType == null) {
+			// Make the current preferred name as non-preferred by setting nstat = 0
+			final Name currentPreferredName = germplasm.findPreferredName();
+			if (currentPreferredName != null) {
+				// nstat = 0 means it is not a preferred name.
+				currentPreferredName.setNstat(0);
+			}
+
+			final Name name = new Name();
+			name.setGermplasm(germplasm);
+			name.setTypeId(nameType.getId());
+			name.setNval(groupName);
+			// nstat = 1 means it is preferred name.
+			name.setNstat(1);
+			// Hard coded to zero for now.
+			name.setLocationId(0);
+			name.setNdate(Util.getCurrentDateAsIntegerValue());
+			name.setReferenceId(0);
+
+			germplasm.getNames().add(name);
+			this.germplasmService.saveGermplasm(germplasm);
```

Comment:
Here maybe use germplasmNamingService 's createName or updateName (to set as non-preferred).

Reply:
I find as a not really good practice to handle the hibernate entity in BMSAPI, it can create some issues when transactions are not correctly configured too. Please use a DTO por Name and as Darla mentioned use GermplasmNamingService